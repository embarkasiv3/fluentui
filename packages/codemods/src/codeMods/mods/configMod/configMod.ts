import { SourceFile } from 'ts-morph';
import {
  CodeMod,
  UpgradeJSONType,
  ModTypes,
  RenamePropModType,
  RepathImportModType,
  CodeModMapType,
} from '../../types';
import { findJsxTag, renameProp, getImportsByPath, repathImport } from '../../utilities/index';

const jsonObj: UpgradeJSONType = require('../upgrades.json');

/* Intermediate file that reads upgrades.json and returns
   a codemod object to be run. */
export function createCodeModFromJson(): CodeMod | undefined {
  return {
    run: (file: SourceFile) => {
      try {
        /* Codemod body, which can be added to */
        const funcs = getCodeModUtilitiesFromJson(file);
        funcs.forEach(func => {
          func();
        });
      } catch (e) {
        return { success: false };
      }
      return { success: true };
    },
    version: '100000',
    name: jsonObj.name,
    enabled: true,
  };
}

/* Helper function that parses a json object for details about individual
   codemods and formats each into a function. These functions are stored in
   an array that is returned to the user. */
export function getCodeModUtilitiesFromJson(file: SourceFile): (() => void)[] {
  const functions = [];
  const modDetails: ModTypes[] = jsonObj.upgrades;
  for (let i = 0; i < modDetails.length; i++) {
    /* Try and get the codemod function associated with the mod type. */
    const func = codeModMap[modDetails[i].type](file, modDetails[i]);
    if (func) {
      functions.push(func);
    } else {
      // eslint-disable-next-line no-throw-literal
      throw 'Error: attempted to access a codeMod mapping from an unsupported type.';
    }
  }
  return functions;
}

/* Dictionary that maps codemod names to functions that execute said mod.
   Used by getCodeModUtilitiesFromJson to easily get the desired function
   from the json object.

   TODO: How well does this scale for devs who want to add mods but don't care about json?*/
const codeModMap: CodeModMapType = {
  renameProp: function(file: SourceFile, mod: RenamePropModType) {
    return function() {
      const tags = findJsxTag(file, mod.options.from.importName);
      renameProp(tags, mod.options.from.toRename, mod.options.to.replacementName);
    };
  },
  repathImport: function(file: SourceFile, mod: RepathImportModType) {
    return function() {
      /* If the json indicates our search string is a regex, convert it. */
      const searchString = mod.options.from.isRegex
        ? new RegExp(
            (mod.options.from.searchString as string)
              .substring(1)
              .substring(0, (mod.options.from.searchString as string).length - 2),
          )
        : mod.options.from.searchString;
      const imports = getImportsByPath(file, searchString);
      imports.forEach(val => {
        repathImport(val, mod.options.to.replacementValue);
      });
    };
  },
};
