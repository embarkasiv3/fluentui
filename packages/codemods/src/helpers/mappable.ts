export interface Mappable<F> {
  fmap: <T>(fn: (v: F) => T) => Mappable<T>;
}
