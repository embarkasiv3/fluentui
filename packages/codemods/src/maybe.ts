/**
 * Maybe is a pattern that helps ensure that you don't need null checks everywhere.
 * I endevoured to make it work the way promises do so it should feel familiar.
 * Using Maybe.then means that you can chain many functions without needing to check for null/undefined
 * constantly.
 *
 * CodeMods is an ideal place to use Maybe because there are numerous cases where something may or may
 * not exist.
 */
export interface MB<T> {
  then: <N>(fn: (v: T) => N | Maybe<N>) => Maybe<N>;
  orElse: (mElse: T) => T;
  __maybe: true;
}

const _makeMaybe = <T>(): MB<T> => {
  const mb: Partial<Maybe<T>> = { __maybe: true };
  mb.then = then.bind(null, mb);
  mb.orElse = orElse.bind(null, mb);
  return mb as MB<T>;
};

export interface Just<T> extends MB<T> {
  value: T;
  just: true;
}

export interface Nothing<T> extends MB<T> {
  just: false;
}

export type Maybe<T> = Just<NonNullable<T>> | Nothing<NonNullable<T>>;

// Need to use assign so that the object remains correctly bound to the methods.
export const Nothing = <T>(): Nothing<T> => Object.assign<MB<T>, { just: false }>(_makeMaybe(), { just: false });

export const Just = <T>(value: NonNullable<T>): Just<NonNullable<T>> => {
  if (value === undefined || value === null) {
    throw new Error('Maybe.Just cannot be undefined or null');
  }
  return Object.assign<MB<NonNullable<T>>, { just: true; value: NonNullable<T> }>(_makeMaybe(), {
    just: true,
    value: value,
  });
};

export const Maybe = <T>(value: T | undefined | null): Maybe<T> => {
  return value !== undefined && value !== null ? Just(value!) : Nothing();
};

/**
 *
 * Works like promises, it will wrap your value in a Maybe if you don't explicitly return one
 *
 */
export const then = <T, N>(mb: Maybe<T>, fn: (v: T) => N | Maybe<N>): Maybe<N> => {
  if (!mb.just) {
    return Nothing();
  }

  const ret = fn(mb.value);
  if ((ret as Maybe<N>).__maybe) {
    return ret as Maybe<N>;
  }
  return Maybe(ret as N);
};

export const orElse = <T>(mb: Maybe<T>, mElse: T): T => {
  return mb.just ? mb.value : mElse;
};
