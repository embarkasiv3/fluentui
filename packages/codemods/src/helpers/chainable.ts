import { Mappable } from './mappable';

export type Flattened<T, V> = T extends Chainable<unknown> ? T : V;
export interface Chainable<T> extends Mappable<T> {
  flatten: () => Flattened<T, Chainable<T>>;
  chain: <R>(fn: (v: T) => Chainable<R>) => Chainable<R>;
  then: <R>(fn: (v: T) => R) => Chainable<R>;
}
