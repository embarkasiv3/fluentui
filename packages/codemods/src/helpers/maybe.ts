import { Chainable, Flattened } from './chainable';

interface MaybeChain<T> extends Chainable<T> {
  something?: boolean;
  __isMaybe: true;
  fmap: <ReturnType>(this: Maybe<T>, fn: (v: NonNullable<T>) => NonNullable<ReturnType>) => Maybe<ReturnType>;
  chain: <ReturnType>(this: Maybe<T>, fn: (v: NonNullable<T>) => Maybe<ReturnType>) => Maybe<ReturnType>;
  flatten: () => Flattened<T, Maybe<T>>;
  then: <ReturnType>(this: Maybe<T>, fn: (v: NonNullable<T>) => ReturnType | Maybe<ReturnType>) => Maybe<ReturnType>;
  orElse: (mElse: NonNullable<T>) => NonNullable<T>;
}
export interface Something<T> extends MaybeChain<T> {
  something: true;
  value: NonNullable<T>;
}
export interface Nothing<T> extends MaybeChain<T> {
  something: false;
}

class MB<T> implements MaybeChain<T> {
  public something: boolean;
  public value: T | undefined;
  public __isMaybe: true = true;

  public fmap<ReturnType>(this: Maybe<T>, fn: (v: NonNullable<T>) => NonNullable<ReturnType>): Maybe<ReturnType> {
    return this.something ? Something(fn(this.value)) : Nothing();
  }

  public chain<R>(this: Maybe<T>, fn: (v: NonNullable<T>) => Maybe<R>): Maybe<R> {
    return this.something ? fn(this.value) : Nothing();
  }

  public flatten() {
    if (this.something) {
      const val = this.value;
      if (typeof val === 'object' && '__isMaybe' in val) {
        return val as Flattened<T, Maybe<T>>;
      }
    }
    // Rewrap the value in a maybe so there is a new object
    return Maybe(this.value) as Flattened<T, Maybe<T>>;
  }

  public then<F>(this: Maybe<T>, fn: (v: NonNullable<T>) => F | Maybe<F>): Maybe<F> {
    return this.chain(v => Maybe(fn(v))).flatten() as Maybe<F>;
  }

  public orElse(mElse: NonNullable<T>): NonNullable<T> {
    return this.something ? this.value! : mElse;
  }
}

export type Maybe<T> = Nothing<T> | Something<T>;

const makeMaybe = <F>(): MaybeChain<F> => {
  return new MB<F>();
};
// Need to use assign so that the object remains correctly bound to the methods.
export const Nothing = <T>(): Nothing<T> => {
  return Object.assign<MaybeChain<T>, { something: false }>(makeMaybe<T>(), { something: false });
};

export const Something = <T>(value: NonNullable<T>): Something<T> => {
  if (value === undefined || value === null) {
    throw new Error('Maybe.Just cannot be undefined or null');
  }
  return Object.assign<MaybeChain<T>, { something: true; value: NonNullable<T> }>(makeMaybe<T>(), {
    something: true,
    value: value,
  });
};

export const Maybe = <T>(value: T | undefined | null): Maybe<T> => {
  return value !== undefined && value !== null ? Something(value!) : Nothing();
};
