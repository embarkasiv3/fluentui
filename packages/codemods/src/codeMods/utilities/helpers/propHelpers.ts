import {
  ts,
  Node,
  SyntaxKind,
  JsxOpeningElement,
  JsxSelfClosingElement,
  VariableDeclarationKind,
  CodeBlockWriter,
  Identifier,
} from 'ts-morph';
import { ValueMap } from 'src/codeMods/types';
import { Maybe } from '../../../helpers/maybe';

/* Helper function to rename a prop if in a spread operator.  */
export function renamePropInSpread(
  element: JsxOpeningElement | JsxSelfClosingElement,
  toRename: string,
  replacementName: string,
  changeValueMap?: ValueMap<string>,
  replacementValue?: string,
) {
  const allAttributes = element.getAttributes();
  allAttributes.forEach(attribute => {
    if (attribute.getKind() === SyntaxKind.JsxSpreadAttribute) {
      /* Get the name of the object that houses the prop. */
      const spreadProp = attribute.getFirstChildByKind(SyntaxKind.Identifier);
      if (spreadContains(toRename, spreadProp)) {
        /* If we've found our prop inside the object,
          continue attempt to rename it. */
        const propKind = spreadProp!.getDefinitions();
        if (propKind.length === 1) {
          const newSpreadName = '__mig' + propKind[0].getName()[0].toUpperCase() + propKind[0].getName().substring(1);
          const newMapName = '__migEnumMap';
          if (verifyVariableDefinition(propKind[0].getKind())) {
            const propContainingObject = propKind[0];
            let newJSXFlag = false;
            const elementType = element.getKind();
            /* Need these data because ELEMENT is invalid if we call propsArrowFunction */
            let blockContainer = getBlockContainer(element);
            if (blockContainer === undefined) {
              const containerMaybe = propsArrowFunction(element);
              if (containerMaybe.something) {
                blockContainer = containerMaybe.value;
                newJSXFlag = true;
              } else {
                // eslint-disable-next-line no-throw-literal
                throw 'attempt to create a new block around prop failed.';
              }
            }
            const parentContainer = blockContainer!.getParentIfKind(SyntaxKind.Block);
            if (parentContainer === undefined) {
              // eslint-disable-next-line no-throw-literal
              throw 'unable to get parent container from block';
            }
            //parentContainer.insertStatements;
            const insertIndex = blockContainer!.getChildIndex();
            if (insertIndex === undefined) {
              // eslint-disable-next-line no-throw-literal
              throw 'unable to find child index';
            }
            if (!parentContainer.getVariableStatement(newSpreadName)) {
              parentContainer.insertVariableStatement(insertIndex, {
                declarationKind: VariableDeclarationKind.Const,
                declarations: [
                  {
                    name: `{${toRename}, ...${newSpreadName}}`,
                    initializer: propContainingObject.getName(),
                  },
                ],
              });
            }
            if (changeValueMap && !parentContainer.getVariableStatement(newMapName)) {
              parentContainer.insertVariableStatement(insertIndex, {
                declarationKind: VariableDeclarationKind.Const,
                declarations: [
                  {
                    name: newMapName,
                    initializer: JSON.stringify(changeValueMap),
                  },
                ],
              });
            }
            let attrToRename = attribute;
            /* attribute is an iterator variable in the forEach function. */
            if (newJSXFlag) {
              const newSpreadProp = Maybe(blockContainer!.getFirstDescendantByKind(SyntaxKind.JsxSpreadAttribute));
              const newJSXElem = Maybe(
                blockContainer!.getFirstDescendantByKind(
                  elementType as SyntaxKind.JsxOpeningElement | SyntaxKind.JsxSelfClosingElement,
                ),
              );
              if (newSpreadProp.something && newJSXElem.something) {
                attrToRename = newSpreadProp.value;
                element = newJSXElem.value;
              }
            }
            /* Cannot use spreadProp because the node has been moved in the AST. */
            attrToRename.replaceWithText(`{...${newSpreadName}}`);
            element.addAttribute({
              name: replacementName,
              initializer: changeValueMap
                ? `{${newMapName}[${toRename}]}`
                : replacementValue
                ? `{${replacementValue}}`
                : `{${toRename}}`,
            });
          }
        }
      }
    }
  });
}

/* Helper function that returns TRUE if the supplied spread object
   contains the prop we're looking for. Else returns FALSE. */
function spreadContains(oldPropName: string, spreadProp?: Identifier): boolean {
  return (
    spreadProp !== undefined &&
    spreadProp
      .getType()
      .getProperties()
      .some(name => name.getName() === oldPropName)
  );
}

/* Verifies that the spread object supplied is a variable
  defined by CONST or LET, or that it exists as a parameter or variable. */
function verifyVariableDefinition(kind: ts.ScriptElementKind): boolean {
  return (
    kind === ts.ScriptElementKind.constElement ||
    kind === ts.ScriptElementKind.letElement ||
    kind === ts.ScriptElementKind.variableElement ||
    kind === ts.ScriptElementKind.parameterElement
  );
}

/* This function is called if no block container could be found,
   meaning that the prop is in a function without a block for code,
   indicating a big-arrow function with no body. We will insert that
   block and then continue. */
function propsArrowFunction(element: JsxOpeningElement | JsxSelfClosingElement) {
  return Maybe(
    element.getFirstAncestor(ans => {
      return ans.getKind() === SyntaxKind.ArrowFunction;
    }),
  )
    .then(ans => {
      const oldBody = element.getText();
      const writer = new CodeBlockWriter();
      writer.block(() => {
        writer.write('return ');
        writer.write(oldBody);
      });
      element.replaceWithText(writer.toString());
      return ans.getFirstChildByKind(SyntaxKind.Block);
    })
    .then(block => {
      return block.getFirstDescendant(child => {
        return child.getKind() === SyntaxKind.ReturnStatement;
      });
    });
}

/* Gets the parent that is a direct descendant of a block
   which should allow for better inserting */
function getBlockContainer(node: Node<ts.Node>) {
  return node.getFirstAncestor(ans => {
    const ansPar = ans.getParent();
    return ansPar?.getKind() === SyntaxKind.Block;
  });
}
