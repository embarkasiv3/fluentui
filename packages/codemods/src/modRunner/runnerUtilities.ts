import { CodeMod, CodeModResult } from '../codeMods/types';
import { Glob } from 'glob';
import { Maybe, Nothing, Something } from '../helpers/maybe';
import { Err } from '../helpers/result';
import { Logger } from './logger';

interface Results {
  modName: string;
  logs: (string | undefined)[];
  status: string;
}

export function runMods<T>(
  codeMods: CodeMod<T>[],
  sources: T[],
  onFileComplete: (result: { file: T; resultList: Results[] }) => void,
  loggingCallback?: (result: { mod: CodeMod<T>; file: T; result: CodeModResult }) => void,
) {
  for (const file of sources) {
    const results: Results[] = [];
    for (let i = 0; i < codeMods.length; i++) {
      const mod = codeMods[i];
      const result = runMod(mod, file, loggingCallback).resolve<Results>(
        ok => ({ modName: mod.name, status: 'success', logs: ok.logs }),
        err => ({ modName: mod.name, status: err.reason, logs: [err.log] }),
      );
      results.push(result);
    }
    onFileComplete({ file, resultList: results });
  }
}

function runMod<T>(
  codeMod: CodeMod<T>,
  file: T,
  loggingCallback?: (result: { mod: CodeMod<T>; file: T; result: CodeModResult }) => void,
): CodeModResult {
  let result: CodeModResult;
  try {
    result = codeMod.run(file);
  } catch (e) {
    result = Err({ reason: 'error', log: e });
  }

  if (loggingCallback) {
    loggingCallback({ mod: codeMod, file: file, result: result });
  }
  return result;
}

export function getModsRootPath() {
  // This function always needs to be in a folder that is a sibling
  // of codeMods.
  return `${__dirname}/../codeMods/mods`;
}

export function getModsPattern(includeTs: boolean = false) {
  // For testing
  if (includeTs) {
    return '/**/*.@(js|ts)';
  }

  return '/**/*.mod.js';
}

export function getModsPaths(root: string = getModsRootPath(), modsPath: string = getModsPattern()) {
  const glob = new Glob(modsPath, {
    absolute: false,
    root: root,
    sync: true,
  });
  return glob.found;
}

// Note, this root will be wherever the npx command is run from.
// For now it will need to be run at the root of the project/monorepo

export function getTsConfigs(root: string = process.cwd()) {
  const glob = new Glob('/**/tsconfig.json', {
    absolute: false,
    ignore: ['**/node_modules/**'],
    root: root,
    sync: true,
  });

  return glob.found;
}

export function loadMod(path: string, errorCallback: (e: Error) => void): Maybe<CodeMod> {
  try {
    const mod = require(path).default;
    return Maybe<CodeMod>(mod);
  } catch (e) {
    errorCallback(e);
  }

  return Nothing();
}

export function getEnabledMods(logger: Logger, getPaths = getModsPaths, loadM = loadMod) {
  return getPaths()
    .map(pth => {
      logger.log('fetching codeMod at ', pth);
      return loadM(pth, e => {
        logger.error(e);
      });
    })
    .filter(modEnabled)
    .map(v => v.value);
}

export function modEnabled<T>(mod: Maybe<CodeMod<T>>): mod is Something<CodeMod<T>> {
  return mod.then(v => !!v.enabled).orElse(false);
}
