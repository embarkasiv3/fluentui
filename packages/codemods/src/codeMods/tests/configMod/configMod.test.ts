import { Project } from 'ts-morph';
import { createCodeModsFromJson } from '../../mods/configMod/configMod';
import { Maybe } from '../../../helpers/maybe';
import { runMods } from '../../../modRunner/runnerUtilities';

const buttonPath = '/**/tests/mock/**/button/**/*.tsx';
const dropDownPath = '/**/tests/mock/**/dropdown/**/*.tsx';

const basicFileName = 'mockImports.tsx';
const oldRoot = 'office-ui-fabric-react';

describe('Tests a simple data-driven codeMod', () => {
  let project: Project;

  beforeEach(() => {
    project = new Project();
    project.addSourceFilesAtPaths(`${process.cwd()}${buttonPath}`);
    project.addSourceFilesAtPaths(`${process.cwd()}${dropDownPath}`);
    project.addSourceFilesAtPaths(`${process.cwd()}/**/tests/mock/utils/*.tsx`);
  });

  it('can run all mods in upgrades.json successfully', () => {
    const mods = Maybe(createCodeModsFromJson());
    if (mods.something) {
      runMods(mods.value, project.getSourceFiles(), result => {
        if (result.error) {
          console.error(`Error running mod ${result.mod.name} on file ${result.file.getBaseName()}`, result.error);
        } else {
          console.log(`Upgraded file ${result.file.getBaseName()} with mod ${result.mod.name}`);
        }
      });
    }
    /* Test for renameProp. */
    expect(project.getSourceFile('mCompoundButtonProps.tsx')?.getFullText()).toMatchSnapshot();
    expect(project.getSourceFile('mDropdownProps.tsx')?.getFullText()).toMatchSnapshot();
    /* Test for repathImport. */
    const file = project.getSourceFileOrThrow(basicFileName);
    file.getImportStringLiterals().forEach(val => {
      const impPath = val.getLiteralValue();
      expect(impPath.indexOf(oldRoot)).not.toEqual(0);
    });
  });
});
