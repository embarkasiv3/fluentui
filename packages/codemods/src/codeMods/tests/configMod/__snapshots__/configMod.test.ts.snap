// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests a simple data-driven codeMod can run all mods in upgrades.json successfully 1`] = `
"import * as React from 'react';
import { CompoundButton } from '@fluentui/react';

export class RenderButton extends React.Component<LocalButtonProps> {
  public render(): JSX.Element {
    const { toggled, id, ...restProps } = this.props;
    return (
      <div>
        <CompoundButton {...this.props} {...restProps} id={'d2'} checked={toggled}>
          Button!
        </CompoundButton>
      </div>
    );
  }
}

export interface LocalButtonProps {
  id: string;
  description?: string;
  imageSource: string;
  toggled?: boolean;
}
"
`;

exports[`Tests a simple data-driven codeMod can run all mods in upgrades.json successfully 2`] = `
"import * as React from 'react';
import { Dropdown } from '@fluentui/react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const RenderDropdown = (props: any) => {
  return (
    <div>
      <Dropdown options={[]} dropdownWidth={0} placeHolder={'placeholder!'} disabled={true} />
      <Dropdown options={[]} dropdownWidth={5} placeHolder={'placeholder!'} disabled={false}>
        {' '}
        Woo Hoo!{' '}
      </Dropdown>
    </div>
  );
};
"
`;
